### ISSUE TITLE
Must the "loop" API be given a new state as the first argument?

### ISSUE THREAD

**Issue by bigradish** [redux-loop/redux-loop#138]
Let's take the code on the frontpage as an example:

`import { loop, Cmd } from 'redux-loop';

function initAction(){
    return {
      type: 'INIT'
    };
}

function fetchUser(userId){
    return fetch(`/api/users/${userId}`);
}

function userFetchSuccessfulAction(user){
   return {
      type: 'USER_FETCH_SUCCESSFUL',
      user
   };
}

function userFetchFailedAction(err){
   return {
      type: 'USER_FETCH_ERROR',
      err
   };
}

const initialState = {
  initStarted: false,
  user: null,
  error: null
};

function reducer(state = initialState, action) {
  switch(action.type) {
  case 'INIT':
    return loop(
      {...state, initStarted: true},
      Cmd.run(fetchUser, {
        successActionCreator: userFetchSuccessfulAction,
        faiLActionCreator: userFetchFailedAction,
        args: ['123']
      })
    );

  case 'USER_FETCH_SUCCESSFUL':
    return {...state, user: action.user};
    
  case 'USER_FETCH_FAILED':
    return {...state, error: action.error};
    
  default:
    return state;
  }
}`

If I do not like putting 'initStarted' in the state, may I keep the state intact in the 'INIT' case in the reducer?
In other words, may I use null as the first argument to the loop api function to info redux-loop to just perform the command and keep the state intact when responding to an action?
I think this feature is useful.
Thank you.

**Comment by bdwain** [comment]
you can't use null but you can use the previous state value. That applies to redux in general. When handling an action in a given reducer, if you don't want to change the state, just return the previous state.

**Comment by bigradish** [comment]
@bdwain Thanks for response. 
If the previous state  is used, the code should be:

case 'INIT':
return loop(
state,
Cmd.run(fetchUser, {
successActionCreator: userFetchSuccessfulAction,
faiLActionCreator: userFetchFailedAction,
args: ['123']
})
);

right? Redux may compare the state with the old one to check if they are equal. If the state is large, the overhead may not be ignored. I think it may be better not to let Redux do the checking.

**Comment by bdwain** [comment]
That code is correct. Redux assumes your state objects are immutable and do not change. http://redux.js.org/docs/introduction/ThreePrinciples.html Changes to the state require you to create a new object with the new state. When your components compare your old and new redux state, they don't do a deep equality check. They do a reference check, which is quick, to see if the states are equal.

**Comment by bigradish** [comment]
@bdwain Thank you. Yes, the comparing in Redux is by reference, and the overhead is small.
However, I still think it should be better not to let Redux even know anything when I just want to run a command.

**Comment by bdwain** [comment]
I think it's pretty much accepted practice at this point that to do "nothing" in a reducer you just return the state that was passed to you (because reducers don't "do" anything, they just return the next state.) I suppose we could allow you to just return a Cmd and redux-loop would add back the state if it detected there was no state returned, but that would complicate the library and I think it'd be confusing because it'd be less like vanilla redux. And there would be no performance benefits either.

**Comment by bigradish** [comment]
@bdwain Thank you. There is some overhead in the code of combineReducers.js 
 ( https://github.com/redux-loop/redux-loop/blob/master/modules/combineReducers.js ), though not big.

I think the code below can be added to the 37th line of combineReducers.js to implement the feature I said:

if (nextStateForKey === null) {
return mutator(model, key, previousStateForKey);
}

As the "reduce" function loop (the 28th line) can not be "break"ed, maybe using a "for" loop instead of the "reduce" function is better (just like: https://github.com/reactjs/redux/blob/master/src/combineReducers.js ). So the added code can be:

if (nextStateForKey === null) {
break;
}

This will avoid iterating all the reducerMap, and then improve the performance.

How do you think of this?

**Comment by bdwain** [comment]
the problem with using null for a returned state to indicate no change is that null is a valid state, and you wouldn't be able to return it anymore if it indicated no change instead of changing the state to null. 

i don't think there's any way to avoid running through every reducer that wouldn't break common assumptions about redux's behavior.

**Comment by bigradish** [comment]
@bdwain Redux state is object, not null, and empty state is represented by {}
Wish you can read your combineReducers.js carefully to see if the iteration-all can be avoided.

**Comment by bdwain** [comment]
redux state can be null

```js
function reducer(state = null, action){
  if(action.type === 'init'){
     return {the: 'new state'};
   }

   return state;
}
```

**Comment by bigradish** [comment]
Where is the code from? Why not:
function reducer(state = {}, action){
  if(action.type === 'init'){
     return {the: 'new state'};
   }

   return state;
}

**Comment by bdwain** [comment]
that also works, but someone might want to communicate something different with null than with an empty object. taking the ability to do that away would be changing something fundamental about redux and would be a lot more confusing to newcomers. 

in [redux's tutorial](http://redux.js.org/docs/basics/Reducers.html), they explicitly mention that returning state in the default case is normal behavior for a reducer.

> We return the previous state in the default case. It's important to return the previous state for any unknown action.

**Comment by bigradish** [comment]
OK, if null is useful for state, a special action and be added, for example:

if (action.reduxLoopType === 'ONLY_DO_CMD') {
break;
}

**Comment by bigradish** [comment]
OK, if null is useful for state, a special action be added:
{
type: '...',
reduxLoopType: 'ONLY_DO_CMD',
reduxLoopCmd: 'cmd to run'
}

in the combineReducers.js file, you can just run the command without iterating.

**Comment by bigradish** [comment]
any response?

**Comment by bdwain** [comment]
Without a clear problem to solve I don't think we should be making a change like this (or any changes really). There would be no performance benefits from this and the behavior it would allow is different from conventional redux, while the current implementation allows us to do the same thing in a much more conventional way.

**Comment by bigradish** [comment]
sigh

**Comment by bdwain** [comment]
sorry. redux makes very few decisions for you, but the one decision it did make was that reducers should be pure functions of the form `(oldState, action) => newState` and they should ALWAYS return the new state no matter what. I would only want to break from that if it enabled some new functionality in a way that was clearly better than any alternatives. Since we can already signal no state change by returning oldState as part of the loop, i'd rather not change things.

I'm going to close this issue for now because we don't have anything actionable for now, but we can continue to discuss it here if you want.
